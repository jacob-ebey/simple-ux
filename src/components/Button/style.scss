@import '../variables';

.button {
  display: inline-block;
  border: $border-width solid $color-dark;
  padding: 5px;
  background-color: $color-light;
  color: $color-dark;
  text-decoration: none;
  cursor: pointer;
  user-select: none;

  &-disabled {
    color: $color-gray;
  }

  &:hover, &:focus {
    border: $border-width solid $color-dark;
    background-color: $color-light-darker;
  }

  &:active {
    border-color: $color-lessdark;
    background-color: $color-light-darkest;
  }

  &-primary {
    border: $border-width solid $color-primary;

    &:hover, &:focus {
      border-color: $color-primary-dark;
    }
  }

  &-secondary {
    border: $border-width solid $color-secondary;

    &:hover, &:focus {
      border-color: $color-secondary-dark;
    }
  }

  &-raise, &-raise-primary, &-raise-secondary {
    &:hover, &:focus {
      box-shadow: 0 0.5em 0.5em -0.5em $color-dark;
      transform: translateY(-0.25em);
    }
  }

  &-pulse {
    &:hover, &:focus {
      animation: -pulse 0.5s;
      box-shadow: 0 0 0 0.5em rgba($color-light, 0);
    }
    @keyframes -pulse {
      0% { box-shadow: 0 0 0 0 $color-dark; }
    }
  }

  &-pulse-primary {
    &:hover, &:focus {
      animation: -pulse-primary 0.5s;
      box-shadow: 0 0 0 0.5em rgba($color-light, 0);
    }
    @keyframes -pulse-primary {
      0% { box-shadow: 0 0 0 0 $color-primary; }
    }
  }

  &-pulse-secondary {
    &:hover, &:focus {
      animation: -pulse-secondary 0.5s;
      box-shadow: 0 0 0 0.5em rgba($color-light, 0);
    }
    @keyframes -pulse-secondary {
      0% { box-shadow: 0 0 0 0 $color-secondary; }
    }
  }
}