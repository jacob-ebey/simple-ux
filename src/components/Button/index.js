import { h, Component } from 'preact' // eslint-disable-line no-unused-vars

import { merge, merge2, mod, raise } from '../tools'

import './style.scss'

/**
 * Button Component
 */
export default class Button extends Component {
  render (props) {
    const { animation, color, disabled, onClick, class: className, children, ...rest } = props

    return (
      <a
        {...rest}
        class={
          merge(
            'button',
            mod('button', color),
            mod('button', disabled && 'disabled'),
            mod('button', merge2(animation, color)),
            className
          )
        }
        onClick={onClick && this._handleClick}
      >
        {children}
      </a>
    )
  }

  _handleClick = (ev) => {
    const { onClick, disabled } = this.props

    raise(onClick && !disabled, () => {
      ev.preventDefault()

      onClick(ev)
    })
  }
}
