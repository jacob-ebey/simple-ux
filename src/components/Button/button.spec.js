/* eslint-env jest */
import { h } from 'preact' // eslint-disable-line no-unused-vars
import { shallow } from 'preact-render-spy'

import Button from './'

describe('Button', () => {
  describe('- snapshots -', () => {
    it('can render', () => {
      const wrapper = shallow(<Button />)
      expect(wrapper).toMatchSnapshot()
    })

    it('can add custom class', () => {
      const wrapper = shallow(<Button class='custom' />)
      expect(wrapper).toMatchSnapshot()
    })

    it('can render children', () => {
      const wrapper = shallow(
        <Button>
          <span>Child 1</span>
          <span>Child 2</span>
        </Button>
      )
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('- logic -', () => {
    it('raises onClick', () => {
      const props = {
        onClick: jest.fn()
      }
      const wrapper = shallow(<Button {...props} />)

      wrapper.find('a').simulate('click', {
        preventDefault: jest.fn()
      })

      expect(props.onClick).toHaveBeenCalledTimes(1)
    })

    it('prevents default', () => {
      const props = {
        onClick: jest.fn()
      }
      const wrapper = shallow(<Button {...props} />)

      const ev = {
        preventDefault: jest.fn()
      }
      wrapper.find('a').simulate('click', ev)

      expect(props.onClick).toHaveBeenCalledTimes(1)
      expect(props.onClick).toHaveBeenCalledWith(ev)

      expect(ev.preventDefault).toHaveBeenCalledTimes(1)
    })
  })
})
