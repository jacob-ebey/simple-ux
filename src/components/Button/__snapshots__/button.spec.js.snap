// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button - snapshots - can add custom class 1`] = `
preact-render-spy (1 nodes)
-------
<a class="button custom"></a>

`;

exports[`Button - snapshots - can be disabled 1`] = `
preact-render-spy (1 nodes)
-------
<a class="button button-disabled"></a>

`;

exports[`Button - snapshots - can render 1`] = `
preact-render-spy (1 nodes)
-------
<a class="button"></a>

`;

exports[`Button - snapshots - can render children 1`] = `
preact-render-spy (1 nodes)
-------
<a class="button">
  <span>Child 1</span>
  <span>Child 2</span>
</a>

`;
